src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Formatting/Passes/CSharpFormattingPass.cs

```diff
@@ -1,97 +1,148 @@
+        _logger.LogTestOnly($"Formatted generated C# document:\r\n{formattedCSharpText}");
+
+        // We now have a formatted C# document, and an original document, but we can't just apply the changes to the original
+        // document as they come from very different places. What we want to do is go through each line of the generated document,
+        // take the indentation that is in it, and apply it to the original document, and then take any formatting changes
+        // on that line, and translate them across to the original document.
+        // Essentially each line is split in two, with indentation on the left of the first non-whitespace char, and formatting
+        // changes on the right. Sometimes we need to skip parts of the right (eg, skip the `@` in `@if`), and sometimes we skip
+        // one side entirely.
+
+        using var formattingChanges = new PooledArrayBuilder<TextChange>();
+        var iFormatted = 0;
+        for (var iOriginal = 0; iOriginal < changedText.Lines.Count; iOriginal++, iFormatted++)
         {
-            changedText = changedText.WithChanges(changes);
-            changedContext = await context.WithTextAsync(changedText, cancellationToken).ConfigureAwait(false);
-        }
-
-        cancellationToken.ThrowIfCancellationRequested();
+            var lineInfo = generator.LineInfo[iOriginal];
 
-        // Apply original C# edits
-        var csharpDocument = roslynWorkspaceHelper.CreateCSharpDocument(changedContext.CodeDocument);
-        var csharpChanges = await FormatCSharpAsync(changedContext, roslynWorkspaceHelper.HostWorkspaceServices, csharpDocument, cancellationToken).ConfigureAwait(false);
-        if (csharpChanges.Length > 0)
-        {
-            changedText = changedText.WithChanges(csharpChanges);
-            changedContext = await changedContext.WithTextAsync(changedText, cancellationToken).ConfigureAwait(false);
+            var formattedLine = formattedCSharpText.Lines[iFormatted];
+            if (lineInfo.ProcessIndentation &&
+                formattedLine.GetFirstNonWhitespaceOffset() is { } formattedIndentation)
+            {
+                var originalLine = changedText.Lines[iOriginal];
+                Debug.Assert(originalLine.GetFirstNonWhitespaceOffset().HasValue);
+
+                var originalLineOffset = originalLine.GetFirstNonWhitespaceOffset().GetValueOrDefault();
+
+                // First up, we take the indentation from the formatted file, and add on the Html indentation level from the line info, and
+                // replace whatever was in the original file with it.
+                var htmlIndentString = context.GetIndentationLevelString(lineInfo.HtmlIndentLevel);
+                var indentationString = formattedCSharpText.ToString(new TextSpan(formattedLine.Start, formattedIndentation))
+                    + htmlIndentString
+                    + lineInfo.AdditionalIndentation;
```

Comment:
Perf note: Consider having a table indentation strings for common cases to avoid allocating a new string for each line.